src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java

```diff
@@ -179,6 +212,39 @@ public void launch(SlaveComputer slaveComputer, TaskListener listener) {
     }
   }
 
+  protected String prepareJavaLaunchString(
+      ComputeEngineInstance node,
+      ComputeEngineComputer computer,
+      Connection conn,
+      PrintStream logger,
+      TaskListener listener)
+      throws IOException, InterruptedException {
+    String javaExecPath = node.getJavaExecPathOrDefault();
+    if (!testCommand(
```

Comment:
This method combines three separate concerns:
1. Checking java is installed.
2. Copying the agent jar to the agent.
3. Formatting a java launch string.
Suggest breaking these up into orthogonal methods which get invoked separately from launch()

Reply:
Separately, if the java exec doesn't exist, do we want to continue uploading the agent jar?

Reply:
Separating concerns: done. This increased the duplication so I decided to complete the de-duplication of the launch method.
Also made it so that failing to check the version means we don't continue.